src/openstack_manager.py

```diff
@@ -113,3 +85,196 @@ def list_projects(cloud_config: dict) -> list[Project]:
+            "default": {
+                key: value
+                for key, value in (
+                    ("httpProxy", http_proxy),
+                    ("httpsProxy", https_proxy),
+                    ("noProxy", no_proxy),
+                )
+                if value
+            }
+        }
+    }
+    docker_client_proxy_content = json.dumps(docker_proxy)
+
+    cmd = [
+        "/usr/bin/bash",
+        BUILD_OPENSTACK_IMAGE_SCRIPT_FILENAME,
+        runner_info["download_url"],
+        str(http_proxy),
+        str(https_proxy),
+        str(no_proxy),
+        docker_proxy_service_conf_content,
+        docker_client_proxy_content,
+    ]
+
+    return cmd
+
+
+@dataclass
+class InstanceConfig:
+    """The configuration values for creating a single runner instance.
+
+    Args:
+        name: Name of the image to launch the GitHub runner instance with.
+        labels: The runner instance labels.
+        registration_token: Token for registering the runner on GitHub.
+        github_path: The GitHub repo/org path
+        openstack_image: The Openstack image to use to boot the instance with.
+    """
+
+    name: str
+    labels: Iterable[str]
+    registration_token: str
+    github_path: GithubPath
+    openstack_image: openstack.image.v2.image.Image
+
+
+def build_image(
+    arch: Arch,
+    cloud_config: dict[str, dict],
+    github_client: GithubClient,
+    path: GithubPath,
+    proxies: Optional[ProxyConfig] = None,
+) -> openstack.image.v2.image.Image:
+    """Build and upload an image to OpenStack.
+
+    Args:
+        cloud_config: The cloud configuration to connect OpenStack with.
+        github_client: The Github client to interact with Github API.
+        path: Github organisation or repository path.
+        proxies: HTTP proxy settings.
+
+    Raises:
+        ImageBuildError: If there were errors building/creating the image.
+
+    Returns:
+        The OpenStack image object.
+    """
+    try:
+        runner_application = github_client.get_runner_application(path=path, arch=arch)
+    except RunnerBinaryError as exc:
+        raise ImageBuildError("Failed to fetch runner application.") from exc
+
+    try:
+        execute_command(_build_image_command(runner_application, proxies), check_exit=True)
+    except SubprocessError as exc:
+        raise ImageBuildError("Failed to build image.") from exc
+
+    try:
+        conn = _create_connection(cloud_config)
+        arch = "amd64" if runner_application["architecture"] == "x64" else "arm64"
+        existing_image: openstack.image.v2.image.Image
+        for existing_image in conn.search_images(name_or_id=IMAGE_NAME):
+            # images with same name (different ID) can be created and will error during server
+            # instantiation.
+            if not conn.delete_image(name_or_id=existing_image.id, wait=True):
+                raise ImageBuildError("Failed to delete duplicate image on Openstack.")
+        return conn.create_image(
```

Comment:
We probably shouldn't return a type from the open stack package in a public function